### ISSUE TITLE
Incorrect last-X-weeks/months/years calculation on BigQuery

### ISSUE THREAD

**Issue by willbryant** [metabase/metabase#25080]
**Describe the bug**
In b9cedccc7f8, @camsaul mostly fixed the BigQuery driver so that it could express the date calculations needed to drive X-weeks/months/years-ago calculations. This gets used whenever we create a filter by last X weeks/month/years.

However, this does not work correctly when the reporting timezone is not UTC. We can see why by testing out the parts of the generated SQL. For a "last 2 weeks, grouped by week" question, with the reporting timezone set to 'Pacific/Auckland' and the first day of the week set to Monday, the SQL looks sometime like this:

```
SELECT timestamp_trunc(`bi.mytable`.`event_at`, week(monday)) AS `event_at`, sum(`bi.mytable`.`somefield`) AS `sum`
FROM `bi.mytable`
WHERE (`bi.mytable`.`event_at` >= CAST(datetime_trunc(datetime_add(current_datetime(), INTERVAL -2 week), week(monday)) AS timestamp)
   AND `bi.mytable`.`event_at` < CAST(datetime_trunc(current_datetime(), week(monday)) AS timestamp))
GROUP BY `event_at`
ORDER BY `event_at` ASC
```

Note that the reporting timezone doesn't appear anywhere in the query. Running the expressions in the BigQuery console:

```
SELECT CAST(datetime_trunc(datetime_add(current_datetime(), INTERVAL -2 week), week(monday)) AS timestamp)
```

Produces `2022-08-15 00:00:00 UTC` - so the week cutoff is for UTC, not the local timezone.

The first issue is that the grouping part, `timestamp_trunc(`bi.mytable`.`event_at`, week(monday))`, needs to specify the timezone - `timestamp_trunc(`bi.mytable`.`event_at`, week(monday), 'Pacific/Auckland')` in my case.

The second part, where the relative calculation is performed, is more complicated. Firstly we would need to specify the timezone to `current_datetime()`, ie. `current_datetime('Pacific/Auckland')`. Unfortunately this isn't enough - the timezone still gets dropped, and the timestamp ends up at 00:00:00 UTC rather than 00:00:00 Pacific/Auckland. (I think BigQuery's handling of datetime timezones is poor here.)

Given that `timestamp_trunc` works correctly (if we give it the timezone), one option is to move the cast so that we do only the absolute minimum in converted datetimes, which is the relative time calculation - ie. just the `datetime_add` - then cast, then use `timestamp_trunc` with reporting timezone. This looks to me to be much simpler than trying to fix the timezone on the `datetime_add`/`datetime_trunc`.

In this approach, we don't even need to specify the timezone to `current_datetime` - it's fine to use UTC for the current time and the relative date calculation, as if we then convert it to a timestamp it's UTC anyway, and only then do we run the truncation in the specific reporting timezone.

So my query above would become:

```
SELECT timestamp_trunc(`bi.mytable`.`event_at`, week(monday), 'Pacific/Auckland') AS `event_at`, sum(`bi.mytable`.`somefield`) AS `sum`
FROM `bi.mytable`
WHERE (`bi.mytable`.`event_at` >= timestamp_trunc(CAST(datetime_add(current_datetime(), INTERVAL -2 week) AS timestamp), week(monday), 'Pacific/Auckland')
   AND `bi.mytable`.`event_at` < timestamp_trunc(CAST(current_datetime() AS timestamp), week(monday), 'Pacific/Auckland'))
GROUP BY `event_at`
ORDER BY `event_at` ASC
```

**To Reproduce**
1. Setup up a BigQuery table with a timestamp column
2. Set the reporting timezone to something other than UTC
3. Create a new question summarising some value by the timestamp at week granularity
4. Filter to the last 2 weeks
5. Observe that rows are filtered and grouped based on the UTC week boundaries, not the reporting timezone

**Expected behavior**
Week/month/year grouping should be performed in the reporting timezone for BigQuery.

**Information about your Metabase Installation:**

You can get this information by going to Admin -> Troubleshooting.

Your browser and the version: Chrome 104
Your operating system: macOS 12.3.1
Your databases: BigQuery
Metabase version: v0.44.1
Metabase hosting environment: Docker
Metabase internal database: H2

**Severity**
How severe an issue is this bug to you? Is this annoying, blocking some users, blocking an upgrade or blocking your usage of Metabase entirely?
Blocking use for financials and causing some weird UX behavior.

**Comment by willbryant** [comment]
The special-case code to workaround timestamp_trunc not supporting week/month/year is here:

```
(defmethod sql.qp/add-interval-honeysql-form :bigquery-cloud-sdk
  [_ hsql-form amount unit]
  ;; `timestamp_add()` doesn't support month/quarter/year, so cast it to `datetime` so we can use `datetime_add()`
  ;; instead in those cases.
  (let [hsql-form (cond->> hsql-form
                    (and (= (temporal-type hsql-form) :timestamp)
                         (not (contains? (temporal-type->supported-units :timestamp) unit)))
                    (hx/cast :datetime))]
    (add-interval-form hsql-form amount unit)))
```

Where we can see that it's doing it at the top level of the expression.

**Comment by willbryant** [comment]
The special-case code to workaround timestamp_trunc not supporting week/month/year is here:

```
(defmethod sql.qp/add-interval-honeysql-form :bigquery-cloud-sdk
  [_ hsql-form amount unit]
  ;; `timestamp_add()` doesn't support month/quarter/year, so cast it to `datetime` so we can use `datetime_add()`
  ;; instead in those cases.
  (let [hsql-form (cond->> hsql-form
                    (and (= (temporal-type hsql-form) :timestamp)
                         (not (contains? (temporal-type->supported-units :timestamp) unit)))
                    (hx/cast :datetime))]
    (add-interval-form hsql-form amount unit)))
```

Where we can see that it's doing it at the top level of the expression.

Interestingly, that seems to call `add-interval-form`, which has its own supported units/datetime cast:

```
(defn- add-interval-form [hsql-form amount unit]
  (let [t         (temporal-type hsql-form)
        add-fn    (temporal-type->arithmetic-function t)
        hsql-form (if (or (not add-fn)
                          (and (not (contains? (temporal-type->supported-units t) unit))
                               (contains? (temporal-type->supported-units :datetime) unit)))
                    (->temporal-type :datetime hsql-form)
                    hsql-form)]
    (AddIntervalForm. hsql-form amount unit)))
```

Which was added much more recently in 001055dfb50 by @metamben.

Perhaps it isn't even needed in `sql.qp/add-interval-honeysql-form` any more and we could concentrate the fixes in just `add-interval-form` and `TruncForm`?

**Comment by willbryant** [comment]
I had a bit of a look to see whether this is something an outsider like me could work on. I've made some progress in the commits above; those commits above fix timezone support for `extract` calls used for granularities like "hour of day", and the `timestamp_trunc` calls used for granularities like "day".

But that still leaves the interval calculations for eg. "last 2 weeks". The big surprise for me so far is that the lines of code I posted above are *not* running their special cases, because the `hsql-form` coming in actually shows up as having a `nil` `temporal-type`.

The `temporal-type` on `sql.qp/current-datetime-honeysql-form :bigquery-cloud-sdk` should be `timestamp`. This defaulting is implemented in `CurrentMomentForm`'s `to-sql`, but the `temporal-type` implementation didn't reflect this.

As a consequence, functions like `sql.qp/add-interval-honeysql-form` weren't able to find the temporal-type and run their workaround.

So in `add-interval-form`, `add-fn` ends up `nil`, and the ->temporal-type :datetime case is always used, which turns out to be the real root cause of this ticket.
